src/Moq/InvocationCollection.cs

```diff
@@ -68,6 +76,9 @@ public void Clear()
 				this.invocations = null;
 				this.count = 0;
 				this.capacity = 0;
+
+				this.owner.Setups.UninvokeAll();
+				// ^ TODO: Currently this could cause a deadlock as another lock will be taken inside this one!
```

Comment:
This raises the question of whether Moq should generally perform more coarse-grained locking.

This would probably mean having just one lock object on the `Mock` instance that would be used both during setups and invocations.

Client code should probably perform all setups first, and only then start invoking the mocked object (possibly from several threads). Code that is structured that way already shouldn't be affected by a move to more coarse-grained locking.